[Validator] Add ContainsValidator

renanbr opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        | n/a

```
/** @Assert\Contains("foo") */
/** @Assert\Contains({"foo", "bar"}) */
```

Commit:
Add StringContains and ContainsWord validators

renanbr commented:
Not sure weather `StringContains` is a better name than `Contains`.

We could have an option called `asWord` (i.e. "extraordinary" will not contain "ordinary")

If this idea gets accepted I can also implements `NotContains` (or add an options `inversed`)

Pierstoval commented on src/Symfony/Component/Validator/Constraints/ContainsValidator.php:
```diff
@@ -0,0 +1,64 @@
+            ) {
+                throw new ConstraintDefinitionException('The Contains constraint expects a valid callback');
+            }
+            $expected = $expected();
```
Maybe we can inject the value and the context here?

renanbr replied:
and also append some `payload` from `Contains` ?

Pierstoval commented on src/Symfony/Component/Validator/Constraints/ContainsValidator.php:
```diff
@@ -0,0 +1,64 @@
+        }
+
+        if (!\is_string($value)) {
+            throw new UnexpectedValueException($value, 'string');
```
I think we can support both strings and arrays since one will just use `strpos` and the other `in_array` (I don't think we should support nested containment for arrays, since it's very specific and can be achieved with `Callback` constraints or `IsTrue` on a valid method)

dorumd replied:
@Pierstoval What do you think about a separate constraint validator for `in_array`?

Pierstoval replied:
In this case, we need `StringContains`, `StringNotContains`, `InArray` and `NotInArray`. That's 4 constraints. Isn't it a bot too much to create 8 classes instead of 4? 😄

dorumd replied:
Well, the way you expose it indeed, it makes no sense. I didn't say anything about `StringNotContains` and `NotInArray`, which are irrelevant as an example and I agree, they don't make any sense.
In the past, I had to work with similar validators/constraints and it was always confusing: is `Contains` doing `in_array` checks or not.

Pierstoval replied:
I'm not entirely against the idea of having separated classes 😉 I just thought it would be easier if `Contains` could handle both cases 😄

stof commented on src/Symfony/Component/Validator/Constraints/ContainsValidator.php:
```diff
@@ -0,0 +1,64 @@
+    public function validate($value, Constraint $constraint)
+    {
+        if (!$constraint instanceof Contains) {
+            throw new UnexpectedTypeException($constraint, __NAMESPACE__.'\Contains');
```
use `Contains::class` instead, as this class will never land in a branch needing to be compatible with PHP < 5.5

xabbuh commented:
When just reading the constraint, I think I would rather expect it to work on some kind of collection (e.g. an array) than checking for a substring.

renanbr commented:
Thanks for your feedbacks, here a summary of current possibilities:

- Options
  - `strategy` (`substring` as default, `word` and `wide`)
  - `case_sensitive` (`true` as default)
- Allows array values
  - It will check against each element, in which one fail generates a violation
  - If both `value` and `expected` are array, each value element **MUST** contains all expected elements

```php
/** @Assert\Contains("foo") */

/** @Assert\Contains({"foo", "bar"}) */

// ok against "A FOO TEXT"
/** @Assert\Contains(expected="foo", caseSensitive=false) */

// fails against "an extraordinary world"
/** @Assert\Contains(expected="ordinary", strategy="word") */

// fails against "foobar"
/** @Assert\Contains(expected="foo", strategy="wide") */
```

stof commented:
I would not make a StringContains constraint work on array values. If you want to validate that all items of an array contain that text, you would use `@assert\All()` wrapping it.
Checking that a collection contains a value should be separate from checking that a string contains some substring IMO (and so this constraint should be renamed to be more specific).

stof commented:
And the `wide` strategy does not make sense. We already have `@Assert\EqualTo`

renanbr commented:
Not sure in how to get consensus :)

I implemented then the stuff I agree with:

- Renamed to `StringContains`
  - Clearly it accepts only `string` value as for now
- Created `StringNotContains`
  - Personally I prefer create a generic `Not`, but I follow `NotBlank`, `NotNull`, etc...
- `strategy` option can be `substring` as default or `word`
  - Why not also `start` and `end`?
- `callback` option receives `context` and `payload`
  - Personally, I prefer the inversed order, but it mimics `Callback` constraint

Pierstoval commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -8,7 +8,8 @@ CHANGELOG
  * added option `allowNull` to NotBlank constraint
  * added `Json` constraint
- 
+ * added `Contains` constraint
```
```suggestion
 * added `StringContains` constraint
 * added `StringNotContains` constraint
```

Pierstoval committed:
Update src/Symfony/Component/Validator/CHANGELOG.md

Co-Authored-By: renanbr <renan-sub@protonmail.com>

renanbr commented on src/Symfony/Component/Validator/Constraints/StringContainsValidator.php:
```diff
@@ -0,0 +1,77 @@
+            throw new UnexpectedValueException($value, 'string');
+        }
+
+        if ($constraint->callback) {
```
Implemented as `resolveExpected(AbstractStringContains $constraint): iterable` in `AbstractStringContainsValidator`

renanbr commented on src/Symfony/Component/Validator/Constraints/StringNotContainsValidator.php:
```diff
@@ -0,0 +1,77 @@
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+use Symfony\Component\Validator\Exception\UnexpectedValueException;
+
+class StringNotContainsValidator extends ConstraintValidator
```
I created `AbstractStringContainsValidator` that provides `resolveExpected()` and `buildRegexPatternTemplate()`.

xabbuh commented:
I am not sure the `strategy` option makes much sense. What I rather can imagine to be useful would be an `ignoreCase` option. For the use case of testing whether a string starts or ends with another string I would implement dedicated `StartsWith` and `EndsWith` constraints.

renanbr commented:
@xabbuh `strategy` option currently solves two problems:

- `substring`, which match with the validator name `StringContains`; and
- `word`, which can be replace/implemented in a dedicated `ContainsWord` (actually is the use-case that motivates me to make this PR)

btw this option doesn't cover `StartsWith` neither `EndsWith`

I'm ok with changing `StringContains`, remove `strategy` option.
We may delivere  `StartsWith`, `EndsWith` and `ContainsWord` validators later...

What do you think ?

xabbuh commented:
IMO that makes sense (my comment about `StartsWith` and `EndsWith` was rather related to your comment about the `strategy` option in https://github.com/symfony/symfony/pull/30581#issuecomment-474949344)

kevin-verschaeve commented on src/Symfony/Component/Validator/Constraints/StringContains.php:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * @var string
+     */
+    public $message = 'This does not contain expected value.';
```
Maybe we can provide the expected text in the message and change the it to something like :

`This value does not contain the expected text: "{{ text }}".`

renanbr replied:
There are two reasons why it doesn't expose `text`/`expected` yet:

- `text` accepts an array, a long list may make it unreadable
- `text` may contains sensible/security data

Anyway, I'll try to use your suggestion if `text` has a single item.

kevin-verschaeve commented on src/Symfony/Component/Validator/Constraints/StringNotContains.php:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * @var string
+     */
+    public $message = 'This contains unexpected value.';
```
As of my previous comment, here we can add a message like this :

`This value contains an unexpected text: "{{ text }}".`

kevin-verschaeve commented on src/Symfony/Component/Validator/Constraints/AbstractStringContains.php:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * @var string|array
+     */
+    public $expected;
```
As this class is extended by `StringContains` and `StringNotContains` I don't think `expected` fit well for the `StringNotContains` constraint, it is a bit confusing. Maybe, a name like `text` would be better ?

vudaltsov replied:
Agree. Another option is to add `abstract protected function getText()` and then expose in subclasses public properties with desired names.

renanbr replied:
@vudaltsov that's almost exactly what I'm doing, I'll push this week some work on it. Spoiler: there are in my local change 8 validators (`StringContains`, `ContainsWord`, `StartsWith`, `EndsWith` and related `Not`s) that will use it.

renanbr commented:
Status update.

Constraints:

- `ContainsWord`
- `NotContainsWord`
- `StringContains`
- `StringNotContains`

Options:

- `text` to be compared
- `callback`, an alternative to `text`
- `payload`, validators will send it to `callback`
- `caseSensitive`, defaults to `false`
- `message`

ro0NL commented on src/Symfony/Component/Validator/Constraints/ContainsWordValidator.php:
```diff
@@ -0,0 +1,32 @@
+    protected function doValidate(string $value, iterable $texts, AbstractStringContains $constraint): void
+    {
+        foreach ($texts as $text) {
+            $pattern = sprintf('/\\b%s\\b/%s', preg_quote($text, '/'), $constraint->caseSensitive ? '' : 'i');
```
IMHO this one doesnt add enough value over RegexValidator :/

ro0NL replied:
then again, a complete set of rich constraints (including StartsWith, EndWith) is nice to have. No blocker for me either way.

renanbr replied:
> IMHO this one doesnt add enough value over RegexValidator :/

The point is UX / code readability

```
@Assert\NotContainsWord({"some", "bad", "words"})
```

vs

```
@Assert\Regex("/^(?!.*(some|bad|words)).*$/", match=false)
```

>  then again, a complete set of rich constraints (including StartsWith, EndWith) is nice to have. No blocker for me either way.

:+1: i'll add them

ro0NL commented on src/Symfony/Component/Validator/Constraints/AbstractStringContainsValidator.php:
```diff
@@ -0,0 +1,104 @@
+        if (isset($constraint->callback)) {
+            $text = $this->loadTextFromCallback($constraint);
+        } else {
+            $text = $constraint->text;
```
i'd allow a callable value as well,  and save us an extra option

renanbr replied:
I tend to agree with that, but

(1) having `callback` is the standard that other validators implement, and

(2) an string is, sometimes, also a callable, i.e. `join`, `explode` ...

ro0NL commented on src/Symfony/Component/Validator/Constraints/AbstractStringContains.php:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * @var string|array|null
+     */
+    public $text;
```
should we really support array (or iterable ;)) values? compared to using the constraint many times ...

ro0NL commented on src/Symfony/Component/Validator/Constraints/StringContainsValidator.php:
```diff
@@ -0,0 +1,32 @@
+    {
+        foreach ($texts as $text) {
+            $pattern = sprintf('/%s/%s', preg_quote($text, '/'), $constraint->caseSensitive ? '' : 'i');
+            if (1 !== preg_match($pattern, $value)) {
```
`strpos / stripos` ?

vudaltsov commented:
> Constraints:
> - ContainsWord
> - NotContainsWord
> - StringContains
> - StringNotContains

Suddenly noticed that we have `EqualTo` without "s", while you propose Contain**s**.
But as far as I understand equal here is not a verb but an adjective with omitted "is". So it's probably okay.

fabpot commented:
@renanbr Can you give us some real-world use cases?

renanbr commented:
My real use case was to check if a text contains dirt words.

Then for others, i.e. `startWith`, we could check if an `Iban` belongs to an specifc country.

The callback could make these validator dynamique by context (session, user preferences, etc).

fabpot commented:
As the list of dirty words is probably coming from some sort of storage, you will have to use a callback, so why not use a callback validator in the first place? This is quite straightforward and it avoids adding many new constraints for which I fail too see use cases without the callback (but I might be wrong).

renanbr commented:
> [...] why not use a callback validator in the first place? This is quite straightforward and it avoids adding many new constraints [...]

IMO it's easer\* to plug a list of dirty words to a native validator than implement it, because regex is hard :-)

\* easer = less code app side, test, etc

> I fail too see use cases without the callback

me too

fabpot commented:
Let's close then.
